# Importing the Python Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Reading the titanic.csv file 

In [2]:
df=pd.read_csv('titanic.csv')
df

pclass                                             name     sex  \
0          1                    Allen, Miss. Elisabeth Walton  female   
1          1                   Allison, Master. Hudson Trevor    male   
2          1                     Allison, Miss. Helen Loraine  female   
3          1             Allison, Mr. Hudson Joshua Creighton    male   
4          1  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   
...      ...                                              ...     ...   
1304       3                             Zabour, Miss. Hileni  female   
1305       3                            Zabour, Miss. Thamine  female   
1306       3                        Zakarian, Mr. Mapriededer    male   
1307       3                              Zakarian, Mr. Ortin    male   
1308       3                               Zimmerman, Mr. Leo    male   

          age  sibsp  parch  ticket      fare    cabin embarked  survived  
0     29.0000      0      0   24160  211.3375       B5        S         1  
1      0.9167      1      2  113781  151.5500  C22 C26        S         1  
2      2.0000      1      2  113781  151.5500  C22 C26        S         0  
3     30.0000      1      2  113781  151.5500  C22 C26        S         0  
4     25.0000      1      2  113781  151.5500  C22 C26        S         0  
...       ...    ...    ...     ...       ...      ...      ...       ...  
1304  14.5000      1      0    2665   14.4542      NaN        C         0  
1305      NaN      1      0    2665   14.4542      NaN        C         0  
1306  26.5000      0      0    2656    7.2250      NaN        C         0  
1307  27.0000      0      0    2670    7.2250      NaN        C         0  
1308  29.0000      0      0  315082    7.8750      NaN        S         0  

[1309 rows x 11 columns]

pclass- Passenger class (1st, 2nd, 3rd)

sibsp - of siblings/spouses aboard

parch - of parents/children aboard

embarked - Port of embarkation (C = Cherbourg, Q = Queenstown, S = Southampton)

survived ---> 0- no, 1-yes


# Understanding the Dataset

In [3]:
df.columns

Index(['pclass', 'name', 'sex', 'age', 'sibsp', 'parch', 'ticket', 'fare',
       'cabin', 'embarked', 'survived'],
      dtype='object')

# Data Cleaning

In [4]:
df.duplicated().any()

np.False_

False - no duplicate 
True - Duplicate


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   pclass    1309 non-null   int64  
 1   name      1309 non-null   object 
 2   sex       1309 non-null   object 
 3   age       1046 non-null   float64
 4   sibsp     1309 non-null   int64  
 5   parch     1309 non-null   int64  
 6   ticket    1309 non-null   object 
 7   fare      1308 non-null   float64
 8   cabin     295 non-null    object 
 9   embarked  1307 non-null   object 
 10  survived  1309 non-null   int64  
dtypes: float64(2), int64(4), object(5)
memory usage: 112.6+ KB


Missing values should be handled for :

Total -1309

age - 1046

fare - 1308

cabin - 295

embarked - 1307

In [6]:
age_mean=round(df['age'].mean(),1)
age_mean

np.float64(29.9)

In [7]:
fare_mean=round(df['fare'].mean(),1)
fare_mean

np.float64(33.3)

number - mean

alphabetic - mode

mode ---> return type series, take the 1st series data [0] and fill the missing valuse using fillna()

In [8]:
emb_mode=df['embarked'].mode()[0]
emb_mode

'S'

Fill the missing data

In [9]:
df['age']=df['age'].fillna(age_mean)

In [10]:
df['fare']= df['fare'].fillna(fare_mean)

In [11]:
df['embarked']= df['embarked'].fillna(emb_mode)

Missing data is handled

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   pclass    1309 non-null   int64  
 1   name      1309 non-null   object 
 2   sex       1309 non-null   object 
 3   age       1309 non-null   float64
 4   sibsp     1309 non-null   int64  
 5   parch     1309 non-null   int64  
 6   ticket    1309 non-null   object 
 7   fare      1309 non-null   float64
 8   cabin     295 non-null    object 
 9   embarked  1309 non-null   object 
 10  survived  1309 non-null   int64  
dtypes: float64(2), int64(4), object(5)
memory usage: 112.6+ KB


In [13]:
df=df.dropna(axis=1)

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   pclass    1309 non-null   int64  
 1   name      1309 non-null   object 
 2   sex       1309 non-null   object 
 3   age       1309 non-null   float64
 4   sibsp     1309 non-null   int64  
 5   parch     1309 non-null   int64  
 6   ticket    1309 non-null   object 
 7   fare      1309 non-null   float64
 8   embarked  1309 non-null   object 
 9   survived  1309 non-null   int64  
dtypes: float64(2), int64(4), object(4)
memory usage: 102.4+ KB


                                                Data Transformation                                                              

In [15]:
df=df.rename(columns={'sex':'gender'})
df.columns

Index(['pclass', 'name', 'gender', 'age', 'sibsp', 'parch', 'ticket', 'fare',
       'embarked', 'survived'],
      dtype='object')

In [16]:
df['sibsp']+df['parch']
df['family_count']=df['sibsp']+df['parch']
df

pclass                                             name  gender  \
0          1                    Allen, Miss. Elisabeth Walton  female   
1          1                   Allison, Master. Hudson Trevor    male   
2          1                     Allison, Miss. Helen Loraine  female   
3          1             Allison, Mr. Hudson Joshua Creighton    male   
4          1  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   
...      ...                                              ...     ...   
1304       3                             Zabour, Miss. Hileni  female   
1305       3                            Zabour, Miss. Thamine  female   
1306       3                        Zakarian, Mr. Mapriededer    male   
1307       3                              Zakarian, Mr. Ortin    male   
1308       3                               Zimmerman, Mr. Leo    male   

          age  sibsp  parch  ticket      fare embarked  survived  family_count  
0     29.0000      0      0   24160  211.3375        S         1             0  
1      0.9167      1      2  113781  151.5500        S         1             3  
2      2.0000      1      2  113781  151.5500        S         0             3  
3     30.0000      1      2  113781  151.5500        S         0             3  
4     25.0000      1      2  113781  151.5500        S         0             3  
...       ...    ...    ...     ...       ...      ...       ...           ...  
1304  14.5000      1      0    2665   14.4542        C         0             1  
1305  29.9000      1      0    2665   14.4542        C         0             1  
1306  26.5000      0      0    2656    7.2250        C         0             0  
1307  27.0000      0      0    2670    7.2250        C         0             0  
1308  29.0000      0      0  315082    7.8750        S         0             0  

[1309 rows x 11 columns]

drop sibsp and parch permanently

In [17]:
df=df.drop(columns=['sibsp','parch'])

df.columns

Index(['pclass', 'name', 'gender', 'age', 'ticket', 'fare', 'embarked',
       'survived', 'family_count'],
      dtype='object')

In [18]:
df['embarked'].unique()

array(['S', 'C', 'Q'], dtype=object)

In [19]:
df['embarked']=df['embarked'].replace({'S':'Southampton','C': 'Cherbourg','Q': 'Queenstown'})
df

pclass                                             name  gender  \
0          1                    Allen, Miss. Elisabeth Walton  female   
1          1                   Allison, Master. Hudson Trevor    male   
2          1                     Allison, Miss. Helen Loraine  female   
3          1             Allison, Mr. Hudson Joshua Creighton    male   
4          1  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   
...      ...                                              ...     ...   
1304       3                             Zabour, Miss. Hileni  female   
1305       3                            Zabour, Miss. Thamine  female   
1306       3                        Zakarian, Mr. Mapriededer    male   
1307       3                              Zakarian, Mr. Ortin    male   
1308       3                               Zimmerman, Mr. Leo    male   

          age  ticket      fare     embarked  survived  family_count  
0     29.0000   24160  211.3375  Southampton         1             0  
1      0.9167  113781  151.5500  Southampton         1             3  
2      2.0000  113781  151.5500  Southampton         0             3  
3     30.0000  113781  151.5500  Southampton         0             3  
4     25.0000  113781  151.5500  Southampton         0             3  
...       ...     ...       ...          ...       ...           ...  
1304  14.5000    2665   14.4542    Cherbourg         0             1  
1305  29.9000    2665   14.4542    Cherbourg         0             1  
1306  26.5000    2656    7.2250    Cherbourg         0             0  
1307  27.0000    2670    7.2250    Cherbourg         0             0  
1308  29.0000  315082    7.8750  Southampton         0             0  

[1309 rows x 9 columns]

In [20]:
df.head()

pclass                                             name  gender      age  \
0       1                    Allen, Miss. Elisabeth Walton  female  29.0000   
1       1                   Allison, Master. Hudson Trevor    male   0.9167   
2       1                     Allison, Miss. Helen Loraine  female   2.0000   
3       1             Allison, Mr. Hudson Joshua Creighton    male  30.0000   
4       1  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.0000   

   ticket      fare     embarked  survived  family_count  
0   24160  211.3375  Southampton         1             0  
1  113781  151.5500  Southampton         1             3  
2  113781  151.5500  Southampton         0             3  
3  113781  151.5500  Southampton         0             3  
4  113781  151.5500  Southampton         0             3

Slpitting name as tilte, first name, last name

In [21]:
names= df['name'].str.split(',',expand=True)
names.head()

0                                        1
0    Allen                   Miss. Elisabeth Walton
1  Allison                    Master. Hudson Trevor
2  Allison                      Miss. Helen Loraine
3  Allison              Mr. Hudson Joshua Creighton
4  Allison   Mrs. Hudson J C (Bessie Waldo Daniels)

In [22]:
df['last_name']=names[0]
df[['title','first_name']]=names[1].str.split('.',n=1,expand=True) #expand creates separate columns
df


pclass                                             name  gender  \
0          1                    Allen, Miss. Elisabeth Walton  female   
1          1                   Allison, Master. Hudson Trevor    male   
2          1                     Allison, Miss. Helen Loraine  female   
3          1             Allison, Mr. Hudson Joshua Creighton    male   
4          1  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   
...      ...                                              ...     ...   
1304       3                             Zabour, Miss. Hileni  female   
1305       3                            Zabour, Miss. Thamine  female   
1306       3                        Zakarian, Mr. Mapriededer    male   
1307       3                              Zakarian, Mr. Ortin    male   
1308       3                               Zimmerman, Mr. Leo    male   

          age  ticket      fare     embarked  survived  family_count  \
0     29.0000   24160  211.3375  Southampton         1             0   
1      0.9167  113781  151.5500  Southampton         1             3   
2      2.0000  113781  151.5500  Southampton         0             3   
3     30.0000  113781  151.5500  Southampton         0             3   
4     25.0000  113781  151.5500  Southampton         0             3   
...       ...     ...       ...          ...       ...           ...   
1304  14.5000    2665   14.4542    Cherbourg         0             1   
1305  29.9000    2665   14.4542    Cherbourg         0             1   
1306  26.5000    2656    7.2250    Cherbourg         0             0   
1307  27.0000    2670    7.2250    Cherbourg         0             0   
1308  29.0000  315082    7.8750  Southampton         0             0   

      last_name    title                          first_name  
0         Allen     Miss                    Elisabeth Walton  
1       Allison   Master                       Hudson Trevor  
2       Allison     Miss                       Helen Loraine  
3       Allison       Mr             Hudson Joshua Creighton  
4       Allison      Mrs   Hudson J C (Bessie Waldo Daniels)  
...         ...      ...                                 ...  
1304     Zabour     Miss                              Hileni  
1305     Zabour     Miss                             Thamine  
1306   Zakarian       Mr                         Mapriededer  
1307   Zakarian       Mr                               Ortin  
1308  Zimmerman       Mr                                 Leo  

[1309 rows x 12 columns]

After splitting drop the 'name' column as it is unneccessary for the dataset

In [23]:
df=df.drop(columns=['name'])
df

pclass  gender      age  ticket      fare     embarked  survived  \
0          1  female  29.0000   24160  211.3375  Southampton         1   
1          1    male   0.9167  113781  151.5500  Southampton         1   
2          1  female   2.0000  113781  151.5500  Southampton         0   
3          1    male  30.0000  113781  151.5500  Southampton         0   
4          1  female  25.0000  113781  151.5500  Southampton         0   
...      ...     ...      ...     ...       ...          ...       ...   
1304       3  female  14.5000    2665   14.4542    Cherbourg         0   
1305       3  female  29.9000    2665   14.4542    Cherbourg         0   
1306       3    male  26.5000    2656    7.2250    Cherbourg         0   
1307       3    male  27.0000    2670    7.2250    Cherbourg         0   
1308       3    male  29.0000  315082    7.8750  Southampton         0   

      family_count  last_name    title                          first_name  
0                0      Allen     Miss                    Elisabeth Walton  
1                3    Allison   Master                       Hudson Trevor  
2                3    Allison     Miss                       Helen Loraine  
3                3    Allison       Mr             Hudson Joshua Creighton  
4                3    Allison      Mrs   Hudson J C (Bessie Waldo Daniels)  
...            ...        ...      ...                                 ...  
1304             1     Zabour     Miss                              Hileni  
1305             1     Zabour     Miss                             Thamine  
1306             0   Zakarian       Mr                         Mapriededer  
1307             0   Zakarian       Mr                               Ortin  
1308             0  Zimmerman       Mr                                 Leo  

[1309 rows x 11 columns]

In [24]:
for i in ['age','fare']:
  df[i]=round(df[i],2)

df.head()
  

pclass  gender    age  ticket    fare     embarked  survived  family_count  \
0       1  female  29.00   24160  211.34  Southampton         1             0   
1       1    male   0.92  113781  151.55  Southampton         1             3   
2       1  female   2.00  113781  151.55  Southampton         0             3   
3       1    male  30.00  113781  151.55  Southampton         0             3   
4       1  female  25.00  113781  151.55  Southampton         0             3   

  last_name    title                          first_name  
0     Allen     Miss                    Elisabeth Walton  
1   Allison   Master                       Hudson Trevor  
2   Allison     Miss                       Helen Loraine  
3   Allison       Mr             Hudson Joshua Creighton  
4   Allison      Mrs   Hudson J C (Bessie Waldo Daniels)

Using a cut() method, Create a age group category

In [25]:
df['age_group']=pd.cut(x=df['age'],bins=[0,3,16,25,60,80],labels=['Baby','Child','Teens','Adult','Senior'])

In [26]:
df.columns

Index(['pclass', 'gender', 'age', 'ticket', 'fare', 'embarked', 'survived',
       'family_count', 'last_name', 'title', 'first_name', 'age_group'],
      dtype='object')

In [27]:
df

pclass  gender    age  ticket    fare     embarked  survived  \
0          1  female  29.00   24160  211.34  Southampton         1   
1          1    male   0.92  113781  151.55  Southampton         1   
2          1  female   2.00  113781  151.55  Southampton         0   
3          1    male  30.00  113781  151.55  Southampton         0   
4          1  female  25.00  113781  151.55  Southampton         0   
...      ...     ...    ...     ...     ...          ...       ...   
1304       3  female  14.50    2665   14.45    Cherbourg         0   
1305       3  female  29.90    2665   14.45    Cherbourg         0   
1306       3    male  26.50    2656    7.22    Cherbourg         0   
1307       3    male  27.00    2670    7.22    Cherbourg         0   
1308       3    male  29.00  315082    7.88  Southampton         0   

      family_count  last_name    title                          first_name  \
0                0      Allen     Miss                    Elisabeth Walton   
1                3    Allison   Master                       Hudson Trevor   
2                3    Allison     Miss                       Helen Loraine   
3                3    Allison       Mr             Hudson Joshua Creighton   
4                3    Allison      Mrs   Hudson J C (Bessie Waldo Daniels)   
...            ...        ...      ...                                 ...   
1304             1     Zabour     Miss                              Hileni   
1305             1     Zabour     Miss                             Thamine   
1306             0   Zakarian       Mr                         Mapriededer   
1307             0   Zakarian       Mr                               Ortin   
1308             0  Zimmerman       Mr                                 Leo   

     age_group  
0        Adult  
1         Baby  
2         Baby  
3        Adult  
4        Teens  
...        ...  
1304     Child  
1305     Adult  
1306     Adult  
1307     Adult  
1308     Adult  

[1309 rows x 12 columns]

Shuffling the columns

In [28]:
df.columns[[9,10,8,1,2,11,5,7,0,3,4,6]]

Index(['title', 'first_name', 'last_name', 'gender', 'age', 'age_group',
       'embarked', 'family_count', 'pclass', 'ticket', 'fare', 'survived'],
      dtype='object')

In [29]:
df=df[df.columns[[9,10,8,1,2,11,5,7,0,3,4,6]]]
df.head()

title                          first_name last_name  gender    age  \
0     Miss                    Elisabeth Walton     Allen  female  29.00   
1   Master                       Hudson Trevor   Allison    male   0.92   
2     Miss                       Helen Loraine   Allison  female   2.00   
3       Mr             Hudson Joshua Creighton   Allison    male  30.00   
4      Mrs   Hudson J C (Bessie Waldo Daniels)   Allison  female  25.00   

  age_group     embarked  family_count  pclass  ticket    fare  survived  
0     Adult  Southampton             0       1   24160  211.34         1  
1      Baby  Southampton             3       1  113781  151.55         1  
2      Baby  Southampton             3       1  113781  151.55         0  
3     Adult  Southampton             3       1  113781  151.55         0  
4     Teens  Southampton             3       1  113781  151.55         0

Since the fare can't be 0 we are handing the dataset again.  
we don't know that the 0 fare passengers are actual passengers or crew_members. 

so we are considering the 0 fare passengers as crew_memebers and
find the index of the members and remove them permanently from the data set, 
then give reset_index as we have removed some index from the middele of the dataset.

In [30]:
crew_members=df.query('fare==0')

crew_members.index

Index([   7,   70,  125,  150,  170,  223,  234,  363,  384,  410,  473,  528,
        581,  896,  898,  963, 1254],
      dtype='int64')

In [31]:
df=df.drop(index=crew_members.index)    

df.reset_index(inplace=True)
df

index    title                          first_name  last_name  gender  \
0         0     Miss                    Elisabeth Walton      Allen  female   
1         1   Master                       Hudson Trevor    Allison    male   
2         2     Miss                       Helen Loraine    Allison  female   
3         3       Mr             Hudson Joshua Creighton    Allison    male   
4         4      Mrs   Hudson J C (Bessie Waldo Daniels)    Allison  female   
...     ...      ...                                 ...        ...     ...   
1287   1304     Miss                              Hileni     Zabour  female   
1288   1305     Miss                             Thamine     Zabour  female   
1289   1306       Mr                         Mapriededer   Zakarian    male   
1290   1307       Mr                               Ortin   Zakarian    male   
1291   1308       Mr                                 Leo  Zimmerman    male   

        age age_group     embarked  family_count  pclass  ticket    fare  \
0     29.00     Adult  Southampton             0       1   24160  211.34   
1      0.92      Baby  Southampton             3       1  113781  151.55   
2      2.00      Baby  Southampton             3       1  113781  151.55   
3     30.00     Adult  Southampton             3       1  113781  151.55   
4     25.00     Teens  Southampton             3       1  113781  151.55   
...     ...       ...          ...           ...     ...     ...     ...   
1287  14.50     Child    Cherbourg             1       3    2665   14.45   
1288  29.90     Adult    Cherbourg             1       3    2665   14.45   
1289  26.50     Adult    Cherbourg             0       3    2656    7.22   
1290  27.00     Adult    Cherbourg             0       3    2670    7.22   
1291  29.00     Adult  Southampton             0       3  315082    7.88   

      survived  
0            1  
1            1  
2            0  
3            0  
4            0  
...        ...  
1287         0  
1288         0  
1289         0  
1290         0  
1291         0  

[1292 rows x 13 columns]

# Data Analysis

1. Get the no.of passengers

Shape[0] ----> gives the no.of rows 

In [32]:
df.shape[0]    

1292

In [33]:
len(df)

1292

2. Get the passenger details whose last_name is Allison

In [34]:
df.query("last_name=='Allison'")   

index    title                          first_name last_name  gender  \
1      1   Master                       Hudson Trevor   Allison    male   
2      2     Miss                       Helen Loraine   Allison  female   
3      3       Mr             Hudson Joshua Creighton   Allison    male   
4      4      Mrs   Hudson J C (Bessie Waldo Daniels)   Allison  female   

     age age_group     embarked  family_count  pclass  ticket    fare  \
1   0.92      Baby  Southampton             3       1  113781  151.55   
2   2.00      Baby  Southampton             3       1  113781  151.55   
3  30.00     Adult  Southampton             3       1  113781  151.55   
4  25.00     Teens  Southampton             3       1  113781  151.55   

   survived  
1         1  
2         0  
3         0  
4         0

In [35]:
df.loc[df['last_name']=='Allison']

index    title                          first_name last_name  gender  \
1      1   Master                       Hudson Trevor   Allison    male   
2      2     Miss                       Helen Loraine   Allison  female   
3      3       Mr             Hudson Joshua Creighton   Allison    male   
4      4      Mrs   Hudson J C (Bessie Waldo Daniels)   Allison  female   

     age age_group     embarked  family_count  pclass  ticket    fare  \
1   0.92      Baby  Southampton             3       1  113781  151.55   
2   2.00      Baby  Southampton             3       1  113781  151.55   
3  30.00     Adult  Southampton             3       1  113781  151.55   
4  25.00     Teens  Southampton             3       1  113781  151.55   

   survived  
1         1  
2         0  
3         0  
4         0

3. Get the pclass 1 male passengers data

In [36]:
df.query("pclass==1 and gender=='male'")

index    title                first_name         last_name gender    age  \
1        1   Master             Hudson Trevor           Allison   male   0.92   
3        3       Mr   Hudson Joshua Creighton           Allison   male  30.00   
5        5       Mr                     Harry          Anderson   male  48.00   
8        9       Mr                     Ramon      Artagaveytia   male  71.00   
9       10      Col                John Jacob             Astor   male  47.00   
..     ...      ...                       ...               ...    ...    ...   
309    316       Mr             Charles Duane          Williams   male  51.00   
310    317       Mr         Richard Norris II          Williams   male  21.00   
311    318       Mr          Fletcher Fellows  Williams-Lambert   male  29.90   
313    320       Mr                      Hugh           Woolner   male  29.90   
314    321       Mr                    George            Wright   male  62.00   

    age_group     embarked  family_count  pclass    ticket    fare  survived  
1        Baby  Southampton             3       1    113781  151.55         1  
3       Adult  Southampton             3       1    113781  151.55         0  
5       Adult  Southampton             0       1     19952   26.55         1  
8      Senior    Cherbourg             0       1  PC 17609   49.50         0  
9       Adult    Cherbourg             1       1  PC 17757  227.52         0  
..        ...          ...           ...     ...       ...     ...       ...  
309     Adult    Cherbourg             1       1  PC 17597   61.38         0  
310     Teens    Cherbourg             1       1  PC 17597   61.38         1  
311     Adult  Southampton             0       1    113510   35.00         0  
313     Adult  Southampton             0       1     19947   35.50         1  
314    Senior  Southampton             0       1    113807   26.55         0  

[172 rows x 13 columns]

4. Get the lowest fare amount

In [37]:
df['fare'].min()

np.float64(3.17)

5. Get the first name and age_category of survived passengers

In [38]:
df.query("survived==0")[['first_name','age_group']]   

first_name age_group
2                          Helen Loraine      Baby
3                Hudson Joshua Creighton     Adult
4      Hudson J C (Bessie Waldo Daniels)     Teens
8                                  Ramon    Senior
9                             John Jacob     Adult
...                                  ...       ...
1287                              Hileni     Child
1288                             Thamine     Adult
1289                         Mapriededer     Adult
1290                               Ortin     Adult
1291                                 Leo     Adult

[794 rows x 2 columns]

6. Get the count of passengers in each gender

In [39]:
g=df.groupby('gender')
g['gender'].count()  

gender
female    466
male      826
Name: gender, dtype: int64

In [40]:
df['gender'].value_counts()

gender
male      826
female    466
Name: count, dtype: int64

7. Get the total fare in each ticket number

In [41]:
g=df.groupby('ticket')
g['fare'].sum()

ticket
110152         259.50
110413         238.95
110465         104.00
110469          26.00
110489          26.55
                ...  
W./C. 6608     171.90
W./C. 6609       7.55
W.E.P. 5734    122.36
W/C 14208       10.50
WE/P 5735      142.00
Name: fare, Length: 918, dtype: float64

In [42]:
pd.pivot_table(df,index='ticket',values='fare',aggfunc='sum')     

fare
ticket             
110152       259.50
110413       238.95
110465       104.00
110469        26.00
110489        26.55
...             ...
W./C. 6608   171.90
W./C. 6609     7.55
W.E.P. 5734  122.36
W/C 14208     10.50
WE/P 5735    142.00

[918 rows x 1 columns]

8. Get the highest,lowest and averge fare in each pclass

In [43]:
g1=df.groupby('pclass')
g1.agg({'fare':['max','min','sum','mean']}) 

fare                           
           max   min       sum       mean
pclass                                   
1       512.33  5.00  28265.42  89.447532
2        73.50  9.69   5866.65  21.648155
3        69.55  3.17   9451.92  13.406979

In [44]:
pd.pivot_table(df,index='pclass',values='fare',aggfunc=['max','min','sum','mean'])

max   min       sum       mean
          fare  fare      fare       fare
pclass                                   
1       512.33  5.00  28265.42  89.447532
2        73.50  9.69   5866.65  21.648155
3        69.55  3.17   9451.92  13.406979

9. Get the total and avg fare,age in each (pclass,gender)

In [45]:
g=df.groupby(['pclass','gender'])
g.agg({'fare':['sum','mean'],'age':['sum','mean']})

fare                   age           
                    sum        mean       sum       mean
pclass gender                                           
1      female  15755.40  109.412500   5254.90  36.492361
       male    12510.02   72.732674   6776.92  39.400698
2      female   2462.89   23.234811   2922.12  27.567170
       male     3403.76   20.628848   5078.13  30.776545
3      female   3310.07   15.324398   5285.77  24.471157
       male     6141.85   12.560020  13237.43  27.070409

In [46]:
pd.pivot_table(df,index=['pclass','gender'],values=['fare','age'],aggfunc=['sum','mean'])

sum                 mean            
                    age      fare        age        fare
pclass gender                                           
1      female   5254.90  15755.40  36.492361  109.412500
       male     6776.92  12510.02  39.400698   72.732674
2      female   2922.12   2462.89  27.567170   23.234811
       male     5078.13   3403.76  30.776545   20.628848
3      female   5285.77   3310.07  24.471157   15.324398
       male    13237.43   6141.85  27.070409   12.560020

10.get the total fare and avg age in each embarked

In [47]:
g=df.groupby(['embarked'])
g.agg({'fare':'sum','age':'mean'})

fare        age
embarked                        
Cherbourg    16830.74  31.809704
Queenstown    1526.27  29.383740
Southampton  25226.98  29.313070

In [48]:
df.query("ticket=='24160'")

index  title                                 first_name last_name  \
0        0   Miss                           Elisabeth Walton     Allen   
175    180   Miss                                     Emilie  Kreuchen   
188    193   Miss                        Georgette Alexandra    Madill   
231    238    Mrs   Edward Scott (Elisabeth Walton McMillan)    Robert   

     gender   age age_group     embarked  family_count  pclass ticket    fare  \
0    female  29.0     Adult  Southampton             0       1  24160  211.34   
175  female  39.0     Adult  Southampton             0       1  24160  211.34   
188  female  15.0     Child  Southampton             1       1  24160  211.34   
231  female  43.0     Adult  Southampton             1       1  24160  211.34   

     survived  
0           1  
175         1  
188         1  
231         1

In [49]:

g2=df.groupby('ticket')
cnt=g2['ticket'].count()

cnt[cnt==1].index

Index(['110469', '110489', '110564', '111163', '111240', '111320', '111369',
       '111426', '111427', '111428',
       ...
       'STON/O2. 3101283', 'STON/O2. 3101290', 'STON/OQ. 369943', 'SW/PP 751',
       'W./C. 14258', 'W./C. 14260', 'W./C. 14263', 'W./C. 14266',
       'W./C. 6609', 'W/C 14208'],
      dtype='object', name='ticket', length=705)

In [50]:
df

index    title                          first_name  last_name  gender  \
0         0     Miss                    Elisabeth Walton      Allen  female   
1         1   Master                       Hudson Trevor    Allison    male   
2         2     Miss                       Helen Loraine    Allison  female   
3         3       Mr             Hudson Joshua Creighton    Allison    male   
4         4      Mrs   Hudson J C (Bessie Waldo Daniels)    Allison  female   
...     ...      ...                                 ...        ...     ...   
1287   1304     Miss                              Hileni     Zabour  female   
1288   1305     Miss                             Thamine     Zabour  female   
1289   1306       Mr                         Mapriededer   Zakarian    male   
1290   1307       Mr                               Ortin   Zakarian    male   
1291   1308       Mr                                 Leo  Zimmerman    male   

        age age_group     embarked  family_count  pclass  ticket    fare  \
0     29.00     Adult  Southampton             0       1   24160  211.34   
1      0.92      Baby  Southampton             3       1  113781  151.55   
2      2.00      Baby  Southampton             3       1  113781  151.55   
3     30.00     Adult  Southampton             3       1  113781  151.55   
4     25.00     Teens  Southampton             3       1  113781  151.55   
...     ...       ...          ...           ...     ...     ...     ...   
1287  14.50     Child    Cherbourg             1       3    2665   14.45   
1288  29.90     Adult    Cherbourg             1       3    2665   14.45   
1289  26.50     Adult    Cherbourg             0       3    2656    7.22   
1290  27.00     Adult    Cherbourg             0       3    2670    7.22   
1291  29.00     Adult  Southampton             0       3  315082    7.88   

      survived  
0            1  
1            1  
2            0  
3            0  
4            0  
...        ...  
1287         0  
1288         0  
1289         0  
1290         0  
1291         0  

[1292 rows x 13 columns]

11. Get the solo passengers details

ticket of solo tavellers

In [51]:
g2=df.groupby('ticket')
cnt=g2['ticket'].count()

idx=cnt[cnt==1].index
idx

Index(['110469', '110489', '110564', '111163', '111240', '111320', '111369',
       '111426', '111427', '111428',
       ...
       'STON/O2. 3101283', 'STON/O2. 3101290', 'STON/OQ. 369943', 'SW/PP 751',
       'W./C. 14258', 'W./C. 14260', 'W./C. 14263', 'W./C. 14266',
       'W./C. 6609', 'W/C 14208'],
      dtype='object', name='ticket', length=705)

no.of solo travellers

In [52]:
g=df.groupby('ticket')
cnt=g['ticket'].count()
cnt[cnt==1].count()

np.int64(705)

# or

In [53]:
g=df.groupby('ticket')
(g['ticket'].count()==1).sum()

np.int64(705)

details of solo travellers

In [54]:
df.loc[df['ticket'].isin(idx)]

index title              first_name     last_name gender   age  \
5         5    Mr                   Harry      Anderson   male  48.0   
8         9    Mr                   Ramon  Artagaveytia   male  71.0   
13       14    Mr   Algernon Henry Wilson     Barkworth   male  80.0   
14       15    Mr                  John D       Baumann   male  29.9   
21       22    Mr             Karl Howell          Behr   male  26.0   
...     ...   ...                     ...           ...    ...   ...   
1284   1301    Mr                 Gerious       Youseff   male  45.5   
1285   1302    Mr                   Wazli        Yousif   male  29.9   
1289   1306    Mr             Mapriededer      Zakarian   male  26.5   
1290   1307    Mr                   Ortin      Zakarian   male  27.0   
1291   1308    Mr                     Leo     Zimmerman   male  29.0   

     age_group     embarked  family_count  pclass    ticket   fare  survived  
5        Adult  Southampton             0       1     19952  26.55         1  
8       Senior    Cherbourg             0       1  PC 17609  49.50         0  
13      Senior  Southampton             0       1     27042  30.00         1  
14       Adult  Southampton             0       1  PC 17318  25.92         0  
21       Adult    Cherbourg             0       1    111369  30.00         1  
...        ...          ...           ...     ...       ...    ...       ...  
1284     Adult    Cherbourg             0       3      2628   7.22         0  
1285     Adult    Cherbourg             0       3      2647   7.22         0  
1289     Adult    Cherbourg             0       3      2656   7.22         0  
1290     Adult    Cherbourg             0       3      2670   7.22         0  
1291     Adult  Southampton             0       3    315082   7.88         0  

[705 rows x 13 columns]

12. Get the no.of passengers in each age category


In [55]:
df.groupby('age_group')['age_group'].count()

C:\Users\Divya Gowda\AppData\Local\Temp\ipykernel_2128\247678786.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('age_group')['age_group'].count()


age_group
Baby       41
Child      93
Teens     307
Adult     818
Senior     33
Name: age_group, dtype: int64

In [56]:
df['age_group'].value_counts()

age_group
Adult     818
Teens     307
Child      93
Baby       41
Senior     33
Name: count, dtype: int64

13. get the age category in which more passengers are present

In [57]:
g=df['age_group']
g.value_counts().idxmax()

'Adult'

14.get the pclass which lowest total fare

total fare in each pclass

In [58]:
df.groupby('pclass')['fare'].sum()


pclass
1    28265.42
2     5866.65
3     9451.92
Name: fare, dtype: float64

OR

In [59]:
pd.pivot_table(df,index='pclass',values='fare',aggfunc='sum')

fare
pclass          
1       28265.42
2        5866.65
3        9451.92

In [60]:

df.groupby('pclass')['fare'].sum().idxmin()

np.int64(2)

15.get the count of survived and non-survived passengers

In [61]:
g=df.groupby('survived')
g['survived'].count()

survived
0    794
1    498
Name: survived, dtype: int64

OR

In [62]:
df['survived'].value_counts()

survived
0    794
1    498
Name: count, dtype: int64

In [63]:
df.groupby('survived').size()

survived
0    794
1    498
dtype: int64

16. get the survived passengers count

In [64]:
g=df.query("survived==1")
g['survived'].count()

np.int64(498)

17. get the gender-wise count of solo travellers

In [65]:
g=df.groupby('gender')
g['gender'].count()

gender
female    466
male      826
Name: gender, dtype: int64

In [66]:
g=df.groupby('ticket')
cnt=g['ticket'].count()
solo_traveller=cnt[cnt==1].index
solo_traveller

Index(['110469', '110489', '110564', '111163', '111240', '111320', '111369',
       '111426', '111427', '111428',
       ...
       'STON/O2. 3101283', 'STON/O2. 3101290', 'STON/OQ. 369943', 'SW/PP 751',
       'W./C. 14258', 'W./C. 14260', 'W./C. 14263', 'W./C. 14266',
       'W./C. 6609', 'W/C 14208'],
      dtype='object', name='ticket', length=705)

In [67]:
df.loc[df['ticket'].isin(solo_traveller)]['gender'].value_counts()

gender
male      546
female    159
Name: count, dtype: int64

OR

In [68]:
g=df[df['ticket'].isin(solo_traveller)]
g['gender'].value_counts()

gender
male      546
female    159
Name: count, dtype: int64